### DISCUSSION TITLE
Starting a recurring function once a particular message has been received.

### DISCUSSION THREAD

**Discussion by sanitgupta** [python-telegram-bot/python-telegram-bot#2470] [Q&A]
Hi,

I recently started using the python-telegram-bot API. I want to write a bot where you initially choose from a bunch of options. Then, depending on which option you choose, a different recurring function is executed say every 10 minutes. I am running it on Heroku using Flask.

I know the JobQueue.run_repeating function is built for exactly this purpose but I am struggling to get it to work in my code currently. For most of my code, I have use functions from the telegram package and not the telegram.ext. I receive texts using:

`update = telegram.Update.de_json(request.get_json(force=True), bot)`

Then, on specific values of the text received, i would like to run/stop  recurring periodic calls of functions. I haven't been able to figure out how to do this here.

Also, a more general questions, what is the purpose behind having these two different packages?

**Comment by Bibo-Joshi** [comment]
Hi. Let me try and answer the questions in reversed order:

> Also, a more general questions, what is the purpose behind having these two different packages?

The `telegram` module provides a pure API wrapper - you'll find everything from core.telegram.org/bots/api in there, some minor auxiliary stuff (like `telegram.constants` and `telegram.error`), but nothing more. For building a bot, this is basically all you need.

The `telegram.ext.` module provides a ton of convenience functionality: The `Updater` class handles the fetching of updates for you. The `Dispatcher` class in combination with the `Handlers` provides a powerful & flexible way to handle updates. You could achieve the same with a lot of `if-else`, sure, but such code gets messy real quick. The idea is basically to provide tools such that you don't need to worry about all the stuff that's common to most bots and instead can focus on when and how your bot is supposed to react.
Because scheduling tasks is a very common use case, we included `JobQueue` into `telegram.ext`, which atm is just a thin wrapper for the `APScheduler` library.

Because `telegram.ext` has some more dependencies (e.g. `APScheduler`) that are not needed when you only need the `telegram` module, we also provide the latter as standalone package via `python-telegram-bot-raw`.

> I know the JobQueue.run_repeating function is built for exactly this purpose but I am struggling to get it to work in my code currently. For most of my code, I have use functions from the telegram package and not the telegram.ext.

Because `JobQueue` needs to run a thread in the background and if you don't use `Updater.start_polling/webhook()`/`Dispatcher.start()`, you need to start it manually via `JobQueue.start()`. Maybe this already helps. If not, please try to specify where you're struggling (I.e. do you get a specific error? Can you build a minimal reproducable example? …)

**Reply by sanitgupta** [reply]
Hi,

Thanks for the answer, I am sharing my code here.

These are the imports I have:
```python
import os
from flask import Flask, request
import telegram
from telegram.ext import Updater, CommandHandler, JobQueue
import requests
from credentials import bot_token, bot_user_name, URL
```

I set a webhook here:
```python
@app.route('/setwebhook', methods = ['GET', 'POST'])
def set_webhook():

	s = bot.setWebhook('{URL}{HOOK}'.format(URL=URL, HOOK=tkn))

	if s:
		return "Webhook OK"
	else:
		return "Webhook FAILED"
```

And then here's my respond function where I have the whole logic for my bot:

```python
@app.route('/{}'.format(tkn), methods = ['POST'])
def respond():

	def callback_text():

		bot.send_message(chat_id = chat_id, text = "sup")


	def reminder():
		bot.send_message(chat_id = chat_id, text = "reminder set")
		JobQueue.start()
		JobQueue.run_repeating(callback = callback_text, interval = 20, context = update)
		
	update = telegram.Update.de_json(request.get_json(force=True), bot)

	chat_id = update.message.chat.id

	msg_id = update.message.message_id

	text = update.message.text.encode('utf-8').decode()

	print("Received message: ", text)

	if text == "random":
		msg = "xyz"
		bot.sendMessage(chat_id = chat_id, reply_to_message_id = msg_id, text = msg)

		return "Sent"

	elif text == "recur":

		msg = "Peridoic Text on"
		bot.sendMessage(chat_id = chat_id, reply_to_message_id = msg_id, text = msg)
		reminder()

		return "Periodic"

	else:
		return "Invalid"
```

Right now, I get the error: JobQueue.start()
`TypeError: start() missing 1 required positional argument: 'self'`

**Reply by Bibo-Joshi** [reply]
I see that you're not using `telegram.ext` at all (Well, you import `Updater` and `CommandHandler`, but don't seem to use them). This is fine, but it also means that using `JobQueue` is of little use for you. The main point of `JobQueue` is to integrate `APScheduler` into the `telegram.ext` framework. I'd recommend to just use `APScheduler` (or any other scheduling library to your liking) directly in this case.

**Reply by sanitgupta** [reply]
Hi, I'll look into other schedulers, thanks.

But for now I was able to get this to work partly and had some more questions. 


Here's what my function looks like:

```python

def respond():

	def callback_text(context: telegram.ext.CallbackContext):

		context.bot.send_message(chat_id = chat_id, text = "sup")
	
	update = telegram.Update.de_json(request.get_json(force=True), bot)

	chat_id = update.message.chat.id

	msg_id = update.message.message_id

	text = update.message.text.encode('utf-8').decode()

	print("Received message: ", text)

	updater = Updater(tkn)
	j = updater.job_queue
	job_min = j.run_repeating(callback_text, interval = 90)
	j.stop()

	if text == "random":
		msg = 'xyz'
		bot.sendMessage(chat_id = chat_id, reply_to_message_id = msg_id, text = msg)
		return "Sent"

	elif text == "rec":
		msg = "Peridoic Text on"
		bot.sendMessage(chat_id = chat_id, reply_to_message_id = msg_id, text = msg)		
		j.start()
		return "Periodic"

	else:
		return "Invalid"
```

This works i.e. when I send the message 'rec' to the bot, it starts a recurring message every 90 seconds.

There are still two problems though: 

1) I expected this to stop the recurring messages as soon as I send any other message to the bot because of the j.stop() before the if condition. That doesn't happen though. e.g. if I send the text 'random' after 'rec', I get an 'xyz' but the recurring message still continues.

2) The heroku app shuts down because of idling even though the job is still running. How can I avoid this?

Thanks!

**Reply by Bibo-Joshi** [reply]
1.: That's not quite how it works. You should define the `JobQueue` (in case you're switching to pure APS the scheduler) *outside* the callback, so only once. Then start the JQ/scheduler and stop it only when your program shuts down. To stop `job_min`, call `job_min.schedule_removal()` (for an APS job it would be `job.remove()`). Ofc you'll have to keep a reference across callbacks for that.

Please note that all this can be made easier for you, if you use the `Dispatcher` class, which can be used with or without threading. See the wiki for more details.

2.: I don't know the details of Heroku, sorry.

**Reply by sanitgupta** [reply]
Hi, thanks again!

I am not sure if I understand. JobQueue is defined outside the callback in my code right? My callback only has one line of code which just sends a message.

Also, I used job_min.schedule_removal() and it didn't stop the job. I want to stop the job even before the program terminates whenever I receive any text which is not 'rec'.

Thanks

**Reply by Bibo-Joshi** [reply]
Well, `JobQueue` is defined within `respond` and if I understand correctly what's going on in flask, that means that you create a new `JobQueue` every time you get a request.

> I want to stop the job even before the program terminates whenever I receive any text which is not 'rec'.

that can't work. How is the job supposed to know when that will be?

But I want to stress again, that it makes no sense to use `JobQueue`, when you're not using `Dispatcher`. I recommend to use plain APScheduler.

**Reply by sanitgupta** [reply]
Ah, I get what you are saying now about the JobQueue creation.

About stopping it, I don't see why it couldn't work. When I receive any text except 'rec', can't I have a if else in my respond function which clears previously created jobs?

I will move to the APScheduler as you have recommended, thanks.

**Reply by Bibo-Joshi** [reply]
> About stopping it, I don't see why it couldn't work. When I receive any text except 'rec', can't I have a if else in my respond function which clears previously created jobs?

Ah, sorry. Misunderstood completely, probably because I'm still looking at the above code, where `j.stop()` stops the `JobQueue`. Then yes, it's perfectly possible to stop a job at runtime. however, you'll have to keep a reference to the current job between calls of  `respond` (or get the reference from the scheduler). In the above snippet you schedule a new job __every__ time `respond` get's called, you If I understand correctly you only want to do that in the `rec` clause.
